### ISSUE TITLE
How to use Bkcrack to hack an MP4 file with store state and zipcrypto?

### ISSUE THREAD

**Issue by Pruserka** [kimci86/bkcrack#130]
Guys! I had a problem where I failed to extract the MP4 video from the file because I didn't know the password. Now the problem is, I don't know everything about this video belt! However, I know exactly what the problem is - I don't know the available plaintext for this MP4. My question is, is there any way to find the available plaintext of this encrypted MP4 file in the store state? I really need your help! Appreciate!

PS: 
```
C:\Users\maggie-yanbo\Desktop>  ./bkcrack -L BC.zip
bkcrack 1.7.0 - 2024-05-26
Archive: BC.zip
Index Encryption Compression CRC32    Uncompressed  Packed size Name
----- ---------- ----------- -------- ------------ ------------ ----------------
    0 None       Store       00000000            0            0 BCm/
    1 ZipCrypto  Deflate     5ce4448f    463567000    462455363 BCm/1.mp4
    2 ZipCrypto  Store       e0a36f9c    257862492    257862504 BCm/10.mp4
    3 ZipCrypto  Store       e41cd6d1    220380314    220380326 BCm/11.mp4
    4 ZipCrypto  Store       274cc0c4    736568233    736568245 BCm/12.mp4
    5 ZipCrypto  Deflate     2ab92a82    450019390    449527646 BCm/2.mp4
    6 ZipCrypto  Deflate     acd5583e    769762919    767931022 BCm/3.mp4
    7 ZipCrypto  Deflate     44157978    467138828    465924987 BCm/4.mp4
    8 ZipCrypto  Deflate     132b2a26    470513700    470028203 BCm/5.mp4
    9 ZipCrypto  Deflate     793fd60a    470807155    470313563 BCm/6.mp4
   10 ZipCrypto  Deflate     cb7801d0    503424017    502233469 BCm/7.mp4
   11 ZipCrypto  Store       f09e01a7    532470572    532470584 BCm/8.mp4
   12 ZipCrypto  Store       41a86692    237160994    237161006 BCm/9.mp4
```

All the passwords in this folder are the same, so I only need to crack a key to solve everything!

**Comment by kimci86** [comment]
Hi, please have a look at my answer on a similar case: https://github.com/kimci86/bkcrack/issues/54#issuecomment-1015376266

**Comment by Pruserka** [comment]
> Hi, please have a look at my answer on a similar case: [#54 (comment)](https://github.com/kimci86/bkcrack/issues/54#issuecomment-1015376266)

Yours means is I should try every codes that you give us, right? Ok, I have a try now! Thanks!

**Comment by Pruserka** [comment]
> Hi, please have a look at my answer on a similar case: [#54 (comment)](https://github.com/kimci86/bkcrack/issues/54#issuecomment-1015376266)

In fact, I have a small question, what plaintext should I use regarding the zip file of this store state?

Index Encryption Compression CRC32    Uncompressed  Packed size Name
----- ---------- ----------- -------- ------------ ------------ ----------------
    0 ZipCrypto  Store       18354f68    325941873    325941885 YLXMm.zip
    1 ZipCrypto  Deflate     8916a7fe        92040        91949 PH.jpg
PS C:\Users\maggie-yanbo\Desktop> ./bkcrack -L YLXM.zip/YLXMm.zip
bkcrack 1.7.0 - 2024-05-26
File error: could not open input file YLXM.zip/YLXMm.zip.

**Comment by kimci86** [comment]
Glad you managed to solve the case with stored mp4. :tada:

> what plaintext should I use regarding the zip file of this store state?

As your encrypted archive contains a stored ZIP file, we can derive some bytes from the ZIP file format specification.
In particular, a ZIP file starts (unless it is empty) with `504b0304` (hexadecimal).
Near the end of a single disk ZIP file of size $n$, assuming there is no archive comment, there is `504b050600000000` starting at offset $n - 22$ and `0000` starting at offset $n - 2$.

For you case (n = 325941873), you can try this:

    ./bkcrack -C YLXM.zip -c YLXMm.zip -x 0 504b0304 -x 325941851 504b050600000000 -x 325941871 0000

Because we provide only 8 contiguous bytes, contrary to the mp4 case, it might take a little while to run.

**Comment by Pruserka** [comment]
> Glad you managed to solve the case with stored mp4. 🎉
> 
> > what plaintext should I use regarding the zip file of this store state?
> 
> As your encrypted archive contains a stored ZIP file, we can derive some bytes from the ZIP file format specification. In particular, a ZIP file starts (unless it is empty) with `504b0304` (hexadecimal). Near the end of a single disk ZIP file of size n, assuming there is no archive comment, there is `504b050600000000` starting at offset n−22 and `0000` starting at offset n−2.
> 
> For you case (n = 325941873), you can try this:
> 
> ```
> ./bkcrack -C YLXM.zip -c YLXMm.zip -x 0 504b0304 -x 325941851 504b050600000000 -x 325941871 0000
> ```
> 
> Because we provide only 8 contiguous bytes, contrary to the mp4 case, it might take a little while to run.

Ok, I got it! Thank you. Now you can close with comment!

**Comment by kimci86** [comment]
So you are saying that you successfully extracted `YLXMm.zip` out of `YLXM.zip`, but `YLXMm.zip` is itself an encrypted archive?
Then we need to guess some bytes again. However, this time, the files are not stored but they are compressed, so we need to guess bytes of compressed data. This is much harder and probably too hard.

Maybe there is another way.
If you can find one of the files in clear somewhere else, then you could try to compress it to get compressed data suitable for a known plaintext attack.
Another idea, maybe this archive is using the same password as one of the other archives you cracked earlier. Then you could try to reuse the corresponding keys to decrypt `YLXMm.zip` and check if the result is valid (i.e. if it can be extracted).
Anyways, you can always attempt a more traditional attack (e.g. brute-force or dictionary) with other tools like as [hashcat](https://hashcat.net/hashcat/) or [john the ripper](https://www.openwall.com/john/).
I hope it helps!

**Comment by Pruserka** [comment]
> So you are saying that you successfully extracted `YLXMm.zip` out of `YLXM.zip`, but `YLXMm.zip` is itself an encrypted archive? Then we need to guess some bytes again. However, this time, the files are not stored but they are compressed, so we need to guess bytes of compressed data. This is much harder and probably too hard.
> 
> Maybe there is another way. If you can find one of the files in clear somewhere else, then you could try to compress it to get compressed data suitable for a known plaintext attack. Another idea, maybe this archive is using the same password as one of the other archives you cracked earlier. Then you could try to reuse the corresponding keys to decrypt `YLXMm.zip` and check if the result is valid (i.e. if it can be extracted). Anyways, you can always attempt a more traditional attack (e.g. brute-force or dictionary) with other tools like as [hashcat](https://hashcat.net/hashcat/) or [john the ripper](https://www.openwall.com/john/). I hope it helps!

Unfortunately, all the methods you gave failed to do. It seems that it is very difficult to hack this file, I decided to give up and now I will close the issue

Thank you very much！Have a good day!

**Comment by nickxyzt** [comment]
I would be very grateful if you could expand on how to find the keys if one has the extracted file from an encrypted zip with compression method "deflate" and encryption method "ZipCrypto".
I tried creating both the encrypted and unencrypted archives, and then find the encryption keys, but I couldn't do it.

Here are the steps I did, please let me know what I'm doing wrong:

# Create the plain text file that gets compressed and encrypted:
echo "This is a plain text file that goes into an archive and must be encrypted" > 1.txt

# Create the encrypted zip file with the password "SimplePass"
zip -r -9 -P SimplePass 1_encrypted.zip 1.txt

# Create also the unencrypted zip file
zip -r -9 1_unencrypted.zip 1.txt

# Show the contents of the archives. Note that the unencrypted file is 12 bytes smaller than the encrypted one (which is normal):

bkcrack -L 1_unencrypted.zip
# Archive: 1_unencrypted.zip
# Index Encryption Compression CRC32    Uncompressed  Packed size Name
# ----- ---------- ----------- -------- ------------ ------------ ----------------
#     0 None       Deflate     4a1b1df9           74           67 1.txt

bkcrack -L 1_encrypted.zip
# Archive: 1_encrypted.zip
# Index Encryption Compression CRC32    Uncompressed  Packed size Name
# ----- ---------- ----------- -------- ------------ ------------ ----------------
#     0 ZipCrypto  Deflate     4a1b1df9           74           79 1.txt


# Inspect the unencrypted zip file, and show bytes in hexadecimal:
tail -c 1000 1_unencrypted.zip | xxd

# Put some characters of the unencrypted compressed file in the plain text file (copy/paste a middle row from the result above):
echo -n "5a310f88af707821944d382be70e5558c07acafe7072c775" | xxd -r -p > plain1.txt

# Try to find the keys:
bkcrack -C 1_encrypted.zip -c 1.txt -p plain1.txt

However, the keys are never found - I've tried in many ways.

Thank you in advance!

**Comment by kimci86** [comment]
By default, bkcrack assumes the given data is the beginning of the plaintext. i.e. the beginning of the original file, or the beginning of the original file after compression if compression was used in the encrypted archive.
Here you are using data from the middle of the compressed file. This is why it fails.

If you want to use the bytes you copied, you have to provide the corresponding offset relative to the beginning of compressed file. Recreating your example locally, I can see you took bytes starting at offset 0x60 in the ZIP file, and the deflate stream starts at offset 0x3f in this ZIP file, so the offset you need to pass is 0x60 - 0x3f = 0x21.

```sh
bkcrack -C 1_encrypted.zip -c 1.txt -p plain1.txt -o 0x21
# or equivalently:
bkcrack -C 1_encrypted.zip -c 1.txt -x 0x21 5a310f88af707821944d382be70e5558c07acafe7072c775
```

Besides, note that copying bytes manually is not needed. You can get bkcrack to load data from the unencrypted archive with the command below. This will also take advantage of having more known plaintext available to speed up the process, compared to having just a small part.

```sh
bkcrack -C 1_encrypted.zip -c 1.txt -P 1_unencrypted.zip -p 1.txt
```

I hope it helps. If you have further questions, please open new issues instead of adding to unrelated existing ones. Thanks!

**Comment by nickxyzt** [comment]
Thank  you very much! I was able to find the password, and used this process to recover the password from the old zip file.
Maybe you should consider putting this in the documentation, as it's quite useful.

Could you also tell how did you identify that the deflated stream starts at offset 0x3f (bytes 0dc9c1...)? Is this a standard header?

**Comment by kimci86** [comment]
I am glad it worked.
The offset option is mentioned vaguely in the [readme](https://github.com/kimci86/bkcrack/blob/master/readme.md#offset), but you are right, the documentation is lacking. I hope to take the time to improve it someday.
To find the offset of the deflate stream in the ZIP file, I opened it in an hexadecimal editor. I know the ZIP format quite well so I simply parsed it by hand. There are tools like [zipdetails](https://github.com/pmqs/zipdetails) to help you doing that. But this work really is not necessary as bkcrack can also parse ZIP files to load entry data from it.
